stl/inc/inplace_vector

```diff
@@ -0,0 +1,1560 @@
+        if (_Count == 0) {
+        } else if (_Count > _N - _My_size) {
+            _Xbad_alloc();
+        } else {
+            const auto _Oldlast           = _My_data + _My_size;
+            const auto _Affected_elements = static_cast<size_type>(_Oldlast - _Whereptr);
+            pointer _Newlast;
+
+            if (_Count < _Affected_elements) { // some affected elements must be assigned
+                _Newlast = _STD uninitialized_move(_Oldlast - _Count, _Oldlast, _Oldlast);
+                _My_size = static_cast<size_type>(_Newlast - _My_data);
+                _STD _Move_backward_unchecked(_Whereptr, _Oldlast - _Count, _Oldlast);
+                _STD _Destroy_range(_Whereptr, _Whereptr + _Count);
+
+                _TRY_BEGIN
+                _STD uninitialized_copy_n(_STD move(_First), _Count, _Whereptr);
+                _CATCH_ALL
+                // glue the broken pieces back together
+
+                _TRY_BEGIN
+                _STD uninitialized_move(_Whereptr + _Count, _Whereptr + 2 * _Count, _Whereptr);
+                _CATCH_ALL
+                // vaporize the detached piece
+                _STD _Destroy_range(_Whereptr + _Count, _Newlast);
+                _My_size = static_cast<size_type>(_Whereptr - _My_data);
+                _RERAISE;
+                _CATCH_END
+
+                _STD _Move_unchecked(_Whereptr + 2 * _Count, _Newlast, _Whereptr + _Count);
+                _STD _Destroy_range(_Oldlast, _Newlast);
+                _My_size = static_cast<size_type>(_Oldlast - _My_data);
+                _RERAISE;
+                _CATCH_END
+            } else { // affected elements don't overlap before/after
+                const pointer _Relocated = _Whereptr + _Count;
+
+                _Newlast = _STD uninitialized_move(_Whereptr, _Oldlast, _Relocated);
+                _My_size = static_cast<size_type>(_Newlast - _My_data);
+                _STD _Destroy_range(_Whereptr, _Oldlast);
+
+                _TRY_BEGIN
+                _STD uninitialized_copy_n(_STD move(_First), _Count, _Whereptr);
+                _CATCH_ALL
+                // glue the broken pieces back together
+
+                _TRY_BEGIN
+                _STD uninitialized_move(_Relocated, _Newlast, _Whereptr);
+                _CATCH_ALL
+                // vaporize the detached piece
+                _STD _Destroy_range(_Relocated, _Newlast);
+                _My_size = static_cast<size_type>(_Whereptr - _My_data);
+                _RERAISE;
+                _CATCH_END
+
+                _STD _Destroy_range(_Relocated, _Newlast);
+                _My_size = static_cast<size_type>(_Newlast - _My_data);
+                _RERAISE;
+                _CATCH_END
+            }
+        }
+    }
+
+    size_type _My_size;
+    union {
+        _Inplace_vector_nontrivial_dummy_type _Dummy;
+        _Ty _My_data[_N];
+    };
+};
+
+template <class _Ty>
+class inplace_vector<_Ty, 0> {

```

Comment:
I think it's better to avoid partial specialization of `inplace_vector` which would make the code harder to maitain.

How about move stored contents into a base class and partially specialize it?